python/pyspark/sql/tests/test_arrow.py

```diff
@@ -191,6 +191,36 @@ def test_pandas_round_trip(self):
         pdf_arrow = df.toPandas()
         assert_frame_equal(pdf_arrow, pdf)
 
+    def test_pandas_self_destruct(self):
+        import pyarrow as pa
+        rows = 2 ** 4
+        cols = 4
+        expected_bytes = rows * cols * 8
+        df = self.spark.range(0, rows).select(*[rand() for _ in range(cols)])
+        # Test the self_destruct behavior by testing _collect_as_arrow directly
+        # Note the sql_conf flags here don't have any real effect as
+        # we aren't going through df.toPandas
+        with self.sql_conf({"spark.sql.execution.arrow.pyspark.selfDestruct.enabled": True}):
+            allocation_before = pa.total_allocated_bytes()
+            batches = df._collect_as_arrow(split_batches=True)
+            table = pa.Table.from_batches(batches)
+            del batches
+            pdf_split = table.to_pandas(self_destruct=True, split_blocks=True, use_threads=False)
+            allocation_after = pa.total_allocated_bytes()
+            difference = allocation_after - allocation_before
+            # Should be around 1x the data size (table should not hold on to any memory)
+            self.assertGreaterEqual(difference, 0.9 * expected_bytes)
+            self.assertLessEqual(difference, 1.1 * expected_bytes)
+        with self.sql_conf({"spark.sql.execution.arrow.pyspark.selfDestruct.enabled": False}):
```

Comment:
This block isn't quite what I was suggesting. We should compare the Pandas DataFrame resulting from `df.toPandas()` with `spark.sql.execution.arrow.pyspark.selfDestruct.enabled` False to `pdf_split`. Also, we will need to call `df.toPandas()` with `spark.sql.execution.arrow.pyspark.selfDestruct.enabled` True and compare that as well.